storage: `CheckSSTConflicts` computes incorrect stats for SST range keys

In #94140, I'm adding kvnemesis test coverage of `AddSSTable` ingestion of MVCC range keys. This found a number of problems:

* Conflicting intents are not handled correctly, errors with `checking for key collisions: not at MVCC value`. Fixed in #94395

* The SST and engine iterators get out of sync, errors with `checking for key collisions: expected engine iter to be ahead of sst iter`. Fixed in #94395

* Incorrect MVCC stats calculations. Not fixed yet

To reproduce, check out the branch in #94140 and run e.g.:

```
$ dev test pkg/kv/kvnemesis -f TestKVNemesisSingleNode -v --stress --ignore-cache
=== RUN   TestKVNemesisSingleNode
I221222 15:43:31.221541 1 (gostd) rand.go:199  [-] 1  random seed: -6583167934786910228
    test_log_scope.go:161: test logs captured to: /tmp/test/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisSingleNode1895708328
    test_log_scope.go:79: use -show-logs to present logs inline
    kvnemesis_test.go:179: seed: 7378274295210697597
    kvnemesis_test.go:123: kvnemesis logging to /tmp/test/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2616062505
    kvnemesis.go:165: error applying x.AddSSTable(ctx, tk(3909762554118877548), tk(18446744073709551615), ... /* @s7 */) // 1221 bytes (as writes)
        // ^-- [tk(7945596544032338278), tk(12435402613537045058)) -> sv(s7): /Table/100/"{6e446df58f9e7166"-ac936cca9ad6fa42"} -> /<empty>
        // ^-- [tk(17786095308637885568), tk(18446744073709551615)) -> sv(s7): /Table/100/"f{6d4e772cc9a5480"-fffffffffffffff"} -> /<empty>: checking for key collisions: expected engine iter to be ahead of sst iter
    kvnemesis.go:185: failures(verbose): /tmp/test/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2616062505/failures
        repro steps: /tmp/test/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2616062505/repro.go
        rangefeed KVs: /tmp/test/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2616062505/kvs-rangefeed.txt
        scan KVs: /tmp/test/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/kvnemesis2616062505/kvs-scan.txt
    kvnemesis_test.go:206: 
        	Error Trace:	/home/erik/.cache/bazel/_bazel_erik/58b8816bb117734f0a0eaaef7fcef95b/sandbox/linux-sandbox/291/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvnemesis/kvnemesis_test_/kvnemesis_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvnemesis/kvnemesis_test.go:206
        	            				/home/erik/.cache/bazel/_bazel_erik/58b8816bb117734f0a0eaaef7fcef95b/sandbox/linux-sandbox/291/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvnemesis/kvnemesis_test_/kvnemesis_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvnemesis/kvnemesis_test.go:147
        	Error:      	Should be zero, but was 1
        	Test:       	TestKVNemesisSingleNode
        	Messages:   	kvnemesis detected failures
    panic.go:522: -- test log scope end --
test logs left over in: /tmp/test/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestKVNemesisSingleNode1895708328
--- FAIL: TestKVNemesisSingleNode (1.58s)
FAIL
I221222 15:43:32.816153 1 (gostd) testmain.go:100  [-] 1  Test //pkg/kv/kvnemesis:kvnemesis_test exited with error code 1


ERROR: exit status 1

11 runs completed, 1 failures, over 2s
```

Jira issue: CRDB-22707


Epic CRDB-20465

--- a/pkg/kv/kvnemesis/generator.go
+++ b/pkg/kv/kvnemesis/generator.go
@@ -525,12 +525,7 @@ func randAddSSTable(g *generator, rng *rand.Rand) Operation {
 	probTombstone := 0.2             // probability to write a tombstone
 	asWrites := rng.Float64() < 0.2  // IngestAsWrites
 
-	if true {
-		// TODO(erikgrinaker): Disable range keys for now since CheckSSTConflicts
-		// computes incorrect MVCC stats. See:
-		// https://github.com/cockroachdb/cockroach/issues/98473
-		numRangeKeys = 0
-	} else if r := rng.Float64(); r < 0.8 {
+	if r := rng.Float64(); r < 0.8 {
 		// 80% probability of only point keys.
 		numRangeKeys = 0
 	} else if r < 0.9 {

--- a/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go
+++ b/pkg/kv/kvserver/batcheval/cmd_add_sstable_test.go
@@ -851,6 +851,20 @@ func TestEvalAddSSTable(t *testing.T) {
 			sst:        kvs{rangeKV("a", "l", 8, "")},
 			expect:     kvs{rangeKV("a", "c", 8, ""), rangeKV("c", "d", 8, ""), rangeKV("c", "d", 6, ""), rangeKV("d", "j", 8, ""), rangeKV("j", "k", 8, ""), rangeKV("j", "k", 5, ""), rangeKV("k", "l", 8, "")},
 		},
+		"DisallowConflicts correctly accounts for complex fragment cases 5": {
+			noConflict: true,
+			reqTS:      10,
+			data:       kvs{pointKV("cc", 7, ""), pointKV("cc", 6, ""), pointKV("cc", 5, "foo"), pointKV("cc", 4, ""), pointKV("cc", 3, "bar"), pointKV("cc", 2, "barfoo"), rangeKV("ab", "g", 1, "")},
+			sst:        kvs{pointKV("aa", 8, "foo"), pointKV("aaa", 8, ""), pointKV("ac", 8, "foo"), rangeKV("b", "c", 8, ""), pointKV("ca", 8, "foo"), pointKV("cb", 8, "foo"), pointKV("cc", 8, "foo"), rangeKV("d", "e", 8, ""), pointKV("e", 8, "foobar")},
+			expect:     kvs{pointKV("aa", 8, "foo"), pointKV("aaa", 8, ""), rangeKV("ab", "b", 1, ""), pointKV("ac", 8, "foo"), rangeKV("b", "c", 8, ""), rangeKV("b", "c", 1, ""), rangeKV("c", "d", 1, ""), pointKV("ca", 8, "foo"), pointKV("cb", 8, "foo"), pointKV("cc", 8, "foo"), pointKV("cc", 7, ""), pointKV("cc", 6, ""), pointKV("cc", 5, "foo"), pointKV("cc", 4, ""), pointKV("cc", 3, "bar"), pointKV("cc", 2, "barfoo"), rangeKV("d", "e", 8, ""), rangeKV("d", "e", 1, ""), rangeKV("e", "g", 1, ""), pointKV("e", 8, "foobar")},
+		},
+		"DisallowConflicts handles existing point key above existing range tombstone": {
+			noConflict: true,
+			reqTS:      10,
+			data:       kvs{pointKV("c", 7, ""), rangeKV("a", "g", 6, ""), pointKV("h", 7, "")},
+			sst:        kvs{rangeKV("b", "d", 8, ""), rangeKV("f", "j", 8, "")},
+			expect:     kvs{rangeKV("a", "b", 6, ""), rangeKV("b", "d", 8, ""), rangeKV("b", "d", 6, ""), pointKV("c", 7, ""), rangeKV("d", "f", 6, ""), rangeKV("f", "g", 8, ""), rangeKV("f", "g", 6, ""), rangeKV("g", "j", 8, ""), pointKV("h", 7, "")},
+		},
 		"DisallowConflicts accounts for point key already deleted in engine": {
 			noConflict: true,
 			reqTS:      10,
@@ -1024,7 +1038,7 @@ func TestEvalAddSSTable(t *testing.T) {
 			noConflict: true,
 			data:       kvs{rangeKV("a", "b", 7, "")},
 			sst:        kvs{rangeKV("a", "b", 7, "")},
-			expectErr:  "ingested range key collides with an existing one",
+			expectErr:  &kvpb.WriteTooOldError{},
 		},
 	}
 	testutils.RunTrueAndFalse(t, "IngestAsWrites", func(t *testing.T, ingestAsWrites bool) {

--- a/pkg/storage/sst.go
+++ b/pkg/storage/sst.go
@@ -473,8 +473,8 @@ func CheckSSTConflicts(
 					isIdempotent := extTombstones.Equal(sstRangeKeys.Versions)
 					if ok := allowIdempotentHelper(extRangeKeys.Versions[0].Timestamp); !ok || !isIdempotent {
 						// Idempotence is either not allowed or there's a conflict.
-						return enginepb.MVCCStats{}, errors.Errorf(
-							"ingested range key collides with an existing one: %s", sstTombstone)
+						return enginepb.MVCCStats{}, kvpb.NewWriteTooOldError(
+							sstTombstone.Timestamp, extRangeKeys.Versions[0].Timestamp.Next(), sstRangeKeys.Bounds.Key)
 					}
 				}
 			}
@@ -813,6 +813,21 @@ func CheckSSTConflicts(
 			// We exclude !sstHasPoint above in case we were at a range key pause
 			// point that matches extKey. In that case, the below SeekGE would make
 			// no forward progress.
+			//
+			// However, seeking is not safe if we're at an ext range key; we could
+			// miss conflicts and overlaps with sst range keys in between
+			// [current SST Key, extKey.Key). Do a next and go back to the start of
+			// the loop. If we had a dedicated sst range key iterator, we could have
+			// optimized away this unconditional-next.
+			if extHasRange {
+				sstIter.NextKey()
+				sstOK, sstErr = sstIter.Valid()
+				if sstOK {
+					extIter.SeekGE(MVCCKey{Key: sstIter.UnsafeKey().Key})
+					extOK, extErr = extIter.Valid()
+				}
+				continue
+			}
 			sstIter.SeekGE(MVCCKey{Key: extKey.Key})
 			sstOK, sstErr = sstIter.Valid()
 			if sstOK {
@@ -1040,10 +1055,10 @@ func CheckSSTConflicts(
 					// ext key.
 					extIter.NextKey()
 					extOK, extErr = extIter.Valid()
-					if extOK {
+					if extOK && extIter.UnsafeKey().Key.Compare(sstIter.UnsafeKey().Key) < 0 {
 						sstIter.SeekGE(MVCCKey{Key: extIter.UnsafeKey().Key})
+						sstOK, sstErr = sstIter.Valid()
 					}
-					sstOK, sstErr = sstIter.Valid()
 					if sstOK {
 						extIter.SeekGE(MVCCKey{Key: sstIter.UnsafeKey().Key})
 						extOK, extErr = extIter.Valid()

